### PR TITLE
tidb-server: Add system time monitor for detecting if system time jum…

### PR DESCRIPTION
**Pull request by ngaut**
…p forward

### CODE REVIEW

File: tidb-server/main.go [hunk 1]
```diff
@@ -112,5 +112,9 @@ func main() {
 		os.Exit(0)
 	}()
 
+	go systimeMonitor(time.Now, func() {
+		log.Fatal("error: system time jump forward")
```
Comment (shenli) [inline tidb-server/main.go]: So after the first time we meet time jump forward, the monitor go routine will exit?

File: tidb-server/systime_mon.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import (
+	"time"
+
+	"github.com/ngaut/log"
+)
+
+type fnTimeNow func() time.Time
+
+// Fatal if system time jump forward
```
Comment (zimulala) [inline tidb-server/systime_mon.go]: Add . at the end of a line

File: tidb-server/systime_mon_test.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package main
+
+import (
+	"testing"
+	"time"
+)
+
+// Fatal if system time jump forward
```
Comment (zimulala) [inline tidb-server/systime_mon_test.go]: The same as here.

**Comment by shenli** [comment]
LGTM

File: tidb-server/main.go [hunk 4]
```diff
@@ -112,5 +112,9 @@ func main() {
 		os.Exit(0)
 	}()
 
+	go systimeMonitor(time.Now, func() {
+		log.Error("error: system time jump forward")
```
Comment (coocood) [inline tidb-server/main.go]: jump backward?

**Comment by ngaut** [comment]
PTAL

**Comment by coocood** [comment]
LGTM

File: tidb-server/main.go [hunk 5]
```diff
@@ -112,5 +113,9 @@ func main() {
 		os.Exit(0)
 	}()
 
+	go systimemon.SystimeMon(time.Now, func() {
```
Comment (zimulala) [inline tidb-server/main.go]: SystimeMon -> SysTimeMon?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**